server/service/src/document/encounter/insert.rs

```diff
@@ -0,0 +1,275 @@
+        };
+        let program_type = "ProgramType".to_string();
+        service_provider
+            .program_service
+            .upsert_program(
+                &ctx,
+                &service_provider,
+                "store_a".to_string(),
+                "user",
+                UpsertProgram {
+                    data: serde_json::to_value(program.clone()).unwrap(),
+                    schema_id: schema.id.clone(),
+                    parent: None,
+                    patient_id: patient.id.clone(),
+                    r#type: program_type.clone(),
+                },
+            )
+            .unwrap();
+
+        // start actual test:
+        let service = &service_provider.encounter_service;
+        // InvalidPatientOrProgram,
+        let err = service
+            .insert_encounter(
+                &ctx,
+                &service_provider,
+                "store_a".to_string(),
+                "user",
+                InsertEncounter {
+                    data: json!({"enrolment_datetime":true}),
+                    schema_id: schema.id.clone(),
+                    patient_id: "some_id".to_string(),
+                    r#type: "SomeType".to_string(),
+                    program_type: program_type.clone(),
+                },
+            )
+            .err()
+            .unwrap();
+        matches!(err, InsertEncounterError::InvalidPatientOrProgram);
+        let err = service
+            .insert_encounter(
+                &ctx,
+                &service_provider,
+                "store_a".to_string(),
+                "user",
+                InsertEncounter {
+                    data: json!({"enrolment_datetime":true}),
+                    schema_id: schema.id.clone(),
+                    patient_id: patient.id.clone(),
+                    r#type: "SomeType".to_string(),
+                    program_type: "invalid".to_string(),
+                },
+            )
+            .err()
+            .unwrap();
+        matches!(err, InsertEncounterError::InvalidPatientOrProgram);
+
+        // InvalidDataSchema
+        let err = service
+            .insert_encounter(
+                &ctx,
+                &service_provider,
+                "store_a".to_string(),
+                "user",
+                InsertEncounter {
+                    data: json!({"encounter_datetime": true}),
+                    schema_id: schema.id.clone(),
+                    patient_id: patient.id.clone(),
+                    r#type: "SomeType".to_string(),
+                    program_type: program_type.clone(),
+                },
+            )
+            .err()
+            .unwrap();
+        matches!(err, InsertEncounterError::InvalidDataSchema(_));
+
+        // success insert
+        let encounter = SchemaEncounter {
+            encounter_datetime: Utc::now().to_rfc3339(),
+            status: None,
+        };
+        let program_type = "ProgramType".to_string();
+        service
```

Comment:
Do we wan to check logic of insert or just database error ?

Reply:
I added a check that the input data matches the data after the update, is that what you meant?

Reply:
yep thanks